src/almond/worker_k8s.ts

```diff
@@ -0,0 +1,304 @@
+        };
+
+        platform.init().then(() => {
+            obj.engine = new Engine(platform, {
+                thingpediaUrl: PlatformModule.thingpediaUrl,
+                nluModelUrl: PlatformModule.nlServerUrl,
+                notifications: PlatformModule.notificationConfig,
+                // nlg will be set to the same URL
+            });
+            if (this.stopped || obj.stopped)
+                return Promise.resolve();
+            return obj.engine.open();
+        }).then(() => {
+            if (this.stopped || obj.stopped)
+                return Promise.resolve();
+
+            obj.running = true;
+            return obj.engine!.run();
+        }).then(() => {
+            return obj.engine!.close();
+        }).catch((e) => {
+            console.error('Engine ' + options.userId + ' had a fatal error: ' + e.message);
+            console.error(e.stack);
+            this.engines.delete(options.userId);
+        });
+
+        this.engines.set(options.userId, obj);
+        return true;
+    }
+
+    killEngine(userId : number) : boolean {
+        console.log(`Killing engine ${userId}`);
+        const obj = this.engines.get(userId);
+        if (!obj)
+            return false;
+        this.engines.delete(userId);
+        obj.stopped = true;
+        if (!obj.engine)
+            return false;
+        obj.engine.stop();
+        for (const sock of obj.sockets)
+            sock.end();
+        return true;
+    }
+
+   engineStatus(userId : number) : string {
+        console.log(`Status of engine ${userId}`);
+        const obj = this.engines.get(userId);
+        if (!obj || !obj.running)
+            return "stopped";
+        return "running";
+    }
+
+    handleDirectSocket(userId : number, replyId : number, socket : stream.Duplex) {
+        console.log('Handling direct connection for ' + userId);
+
+        const rpcSocket = new rpc.Socket(new JsonDatagramSocket(socket, socket, 'utf8'));
+        rpcSocket.on('error', (e) => {
+            console.log('Error on direct RPC socket: ' + e.message);
+        });
+
+        const obj = this.engines.get(userId);
+        if (!obj || !obj.engine) {
+            console.log('Could not find an engine with the required user ID');
+            rpcSocket.call(replyId, 'error', ['Invalid user ID ' + userId]);
+            rpcSocket.end();
+            return;
+        }
+
+        const platform = obj.engine.platform;
+        rpcSocket.call(replyId, 'ready', [
+            obj.engine,
+            platform.getCapability('websocket-api'),
+            platform.getCapability('webhook-api')
+        ]);
+
+        obj.sockets.add(rpcSocket);
+        rpcSocket.on('close', () => {
+            obj.sockets.delete(rpcSocket);
+        });
+    }
+
+    async connectWSEngine(ws : WebSocket) {
+        console.log('connecting engine ...');
+        const socket = wsjs.createWebSocketStream(ws);
+        const jsonSocket = new JsonDatagramSocket(socket, socket, 'utf-8');
```

Comment:
Wait this code doesn't look right: you first convert the WebSocket (which is a datagram transport because websocket messages are framed) into a streaming transport, and then convert it again into a datagram transport with JsonDatagramSocket. I don't think we want to do that: we can directly read/write messages from the WebSocket and pipe into the RPC socket.

Also, the whole initialization dance is unnecessary: there are no more control sockets, because there is no master using transparent-rpc to talk to the worker. All sockets are direct sockets, and we can pass the initialization info as query string or similar, so we can immediately speak the RPC protocol.

Reply:
Please see my latest websocket - transparent rpc changes.  I didn't add authentication yet. Shouldn't that happen at the handshake using control message?